File: tests/core/separators/codac2_tests_SepVisible.cpp
/** * Codac tests - Visibility Separator
 * ----------------------------------------------------------------------------
 * \date       2026
 * \author     Quentin Brateau
 * \copyright  Copyright 2024 Codac Team
 * \license    GNU Lesser General Public License (LGPL)
 */

#include <catch2/catch_test_macros.hpp>
#include <codac2_SepVisible.h>

using namespace codac2;

TEST_CASE("SepVisible - Space Partitioning")
{
  Vector a({0.0, 0.0});
  Segment s({1.0, 1.0}, {1.0, -1.0}); // Vertical wall at x=1
  SepVisible sep(a, s);

  SECTION("Consistency check: in | out should cover boundary")
  {
    IntervalVector x({{0.0, 2.0}, {-2.0, 2.0}});
    BoxPair res = sep.separate(x);
    
    // The union of the contracted boxes should ideally cover the original box 
    // (minus the parts definitively removed by contractors)
    CHECK(!res.inner.is_empty());  // Some parts are hidden (x > 1)
    CHECK(!res.outer.is_empty()); // Some parts are visible (x < 1 or y > cone)
  }

  SECTION("Corner case: Box exactly on the observation point")
  {
    // A point at the source is always visible (or at least not hidden by the obstacle)
    IntervalVector x({{0.0, 0.0}, {0.0, 0.0}});
    BoxPair res = sep.separate(x);
    
    CHECK(res.inner.is_empty());    // Not hidden
    CHECK(res.outer == x);         // Visible
  }
}

File: tests/core/separators/codac2_tests_SepVisible.py
#!/usr/bin/env python

#  Codac tests - Visibility
# ----------------------------------------------------------------------------
#  \date       2026
#  \author     Quentin Brateau
#  \copyright  Copyright 2026 Codac Team
#  \license    GNU Lesser General Public License (LGPL)

import unittest
from codac import *

class TestVisibility(unittest.TestCase):

  def test_SepVisible(self):
    a = [0.0, 0.0]
    s = Segment([1.0, 1.0], [1.0, -1.0])
    sep = SepVisible(a, s)

    # 1. Space Partitioning
    x = IntervalVector([[0.0, 2.0], [-2.0, 2.0]])
    x_in, x_out = sep.separate(x)
    self.assertFalse(x_in.is_empty())
    self.assertFalse(x_out.is_empty())

    # 2. Box exactly on observation point
    x_point = IntervalVector([[0.0, 0.0], [0.0, 0.0]])
    x_in, x_out = sep.separate(x_point)
    self.assertTrue(x_in.is_empty())
    self.assertTrue(x_out == x_point)


if __name__ == '__main__':
  unittest.main()